src/Common/ZooKeeper/ZooKeeperRetries.h

```diff
@@ -221,7 +220,8 @@ class ZooKeeperRetriesControl
         }
 
         if (process_list_element && !process_list_element->checkTimeLimitSoft())
-            return false;
+            throw Exception(
```

Comment:
I would suggest introducing a new function instead of `checkTimeLimitSoft()` that throws in the cases it's ok to throw (KILL or timeout with mode THROW). Otherwise we'd be throwing here with mode BREAK, which is kinda unexpected. It also gives us the option to say exactly what happened with a clear error message

Reply:
> that throws in the cases it's ok to throw (KILL or timeout with mode THROW)

You described `checkTimeLimit` which already exists.
But isn't the point here that we want to ignore timeout mode and throw in every case?

Reply:
Maybe more explicit would be adding function which checks if the limit is exceeded? there already is `isKilled`

Reply:
> You described checkTimeLimit which already exists.

Seems like it is. Let's use it instead.

> But isn't the point here that we want to ignore timeout mode and throw in every case?

No. If the user requested `BREAK` we don't want to throw as it halts the execution (which they've explicitly requested not to do). IMO it's better to continue running in that situation, even if timeouts are not respected.

Reply:
we will still have the same problem
For example in https://github.com/ClickHouse/ClickHouse/issues/60882
if `BREAK` is requested, execution will continue and a `LOGICAL_ERROR` will be thrown.
My thought process was to prefer a better exception even though it ignores timeout mode than throw `LOGICAL_ERROR`.

Reply:
I mean to continue on BREAK (no check), that is keep the loop going and sending keeper requests as if there was no timeout.

Reply:
That makes sense